### PR TITLE
Issue #6805: fix false-positive in while idiom for InnerAssingmentCheck

### PR DESCRIPTION
**Pull request by strkkk**
#6805
Reg report: https://strkkk.github.io/6805/

### CODE REVIEW

**Comment by strkkk** [comment]
Teamcity complains about `SuppressWithPlainTextCommentFilter` and its nested package-private class, but I didn't touch it.
It looks like master is not stable.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java [hunk 1]
```diff
@@ -264,4 +279,19 @@ private static boolean isInContext(DetailAST ast, int[]... contextSet) {
...
+     * Get ast parent, ignoring token types from {@code skipTokens}.
+     *
+     * @param ast token to get parent
+     * @param skipTokens token types to skip
+     * @return first not ignored parent of ast
+     */
+    private static DetailAST getParent(DetailAST ast, int... skipTokens) {
+        DetailAST result = ast.getParent();
+        while (Arrays.binarySearch(skipTokens, result.getType()) > -1) {
+            result = result.getParent();
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: Isn't this bad to be endlessly checking parent after parent?
What happens if we have an anonymous class inside the while loop. If there is an assignment in the anonymous class, won't this break out of the anonymous class and see the while loop and falsely pick that as the parent?
Something to that nature.

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: Can you please provide code example?

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: I am not sure where this code is in terms of the tree it is looking at, but I am thinking of something along the lines:
````
while (true && new Test() {
    public boolean test() {
        boolean result;
        result = true;
        return result;
    }.getResult()) {
  // example
}
````

Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: I still do not understand. 
Before it was looking on direct parent and checked if it matches sequence. Right now it does exactly the same thing for other cases, but for while idiom it skips boolean operators.
In your code there are no boolean operators in assignment parents.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: I will have to play with the code more then I can probably explain better.

**Comment by romani** [comment]
@strkkk , please rebase to fix ci.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java [hunk 2]
```diff
@@ -119,8 +119,20 @@
         TokenTypes.NOT_EQUAL,
     };
 
+    /**
+     * The token types that are ignored while checking "while-idiom".
+     */
+    private static final int[] WHILE_IDIOM_IGNORED_PARENTS = {
```
[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: There is already things like
https://github.com/checkstyle/checkstyle/blob/6ee53318f52712bc8d89dbd3cb748aec958df9fb/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignment.java#L25

and 
https://github.com/checkstyle/checkstyle/blob/6ee53318f52712bc8d89dbd3cb748aec958df9fb/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignment.java#L39-L42

and 
https://github.com/checkstyle/checkstyle/blob/6ee53318f52712bc8d89dbd3cb748aec958df9fb/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignment.java#L65

Please specify what cases you mean. For loops are not affected by this code changes

> can we name it as Loop-idiom ?

Other loops are supported?

[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: @romani ping

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: Sorry fo delay, I am on vacation, so have shortage of time.
FOR loop should be able to have same model of assignment as WHILE. If we support only WHILE is it is better to update doc. If we support both, we still need to update doc to use more general term.

I tried to find time to give you good feedback but failed, so it is is reason of delay.

[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: In my opinion, such corrections is out of scope. It is supposed to be just a correction for while idiom false-positive.

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: No problem, please create outbound issue if you agree that this is a problem.

[the remark this answers is not in the document]
Comment (strkkk) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: I do not sure if this is corrections is needed, but I still think it should be done separately.
I created #6915 discussion can be continued there.

Is there some other review items?

[the remark this answers is not in the document]
Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/InnerAssignmentCheck.java]: No other items, separate issue is good point, I just try to keep track of problems.

**Comment by romani** [comment]
@rnveach , please do second review

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignment.java [hunk 3]
```diff
@@ -111,4 +111,18 @@ public ChildParent getParent() {
...
+        FileInputStream file = null;
+        while (!((line = file.read()) != null)) {}
+        while ((line = file.read()) != null && line < 3) {}
+        while ((line = file.read()) != null && line < 3 && line > 5) {}
+        while ((line = file.read()) != null || line < 3) {}
+        while ((line = file.read()) != null || line < 3 || line > 5) {}
+        while ((line = file.read()) != null & line < 3) {}
+        while ((line = file.read()) != null & line < 3 & line > 5) {}
+        while ((line = file.read()) != null | line < 3) {}
+        while ((line = file.read()) != null | line < 3 | line > 5) {}
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignment.java]: After seeing https://strkkk.github.io/6805/RxJava/xref/checkstyle/contribution/checkstyle-tester/repositories/RxJava/src/main/java/rx/internal/operators/OperatorMergeMaxConcurrent.java.html#L264 , please add an example where assignment isn't first thing in the condition.

Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/innerassignment/InputInnerAssignment.java]: Done

**Comment by strkkk** [comment]
Travis failure is not related with code changes, ci is ok.
> The job exceeded the maximum log length, and has been terminated.

**Pull request merged**